From 579dcdbc8aab966b22f6324a16dbf8fd08a607c3 Mon Sep 17 00:00:00 2001
From: "Anthony G. Basile" <blueness@gentoo.org>
Date: Fri, 17 Apr 2015 16:51:43 -0400
Subject: [PATCH] ldso/libdl/libdl.c: do not unmap libraries.

---
 ldso/libdl/libdl.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ldso/libdl/libdl.c b/ldso/libdl/libdl.c
index 49711a8..2c43c07 100644
--- a/ldso/libdl/libdl.c
+++ b/ldso/libdl/libdl.c
@@ -977,7 +977,7 @@ static int do_dlclose(void *vhandle, int need_fini)
 
 			end = (end + ADDR_ALIGN) & PAGE_ALIGN;
 			start = start & ~ADDR_ALIGN;
-			DL_LIB_UNMAP (tpnt, end - start);
+			//DL_LIB_UNMAP (tpnt, end - start);
 			/* Free elements in RTLD_LOCAL scope list */
 			for (runp = tpnt->rtld_local; runp; runp = tmp) {
 				tmp = runp->next;
-- 
2.0.5

