diff -urN livecd-tools-1.0.33.orig/x-setup livecd-tools-1.0.33/x-setup
--- livecd-tools-1.0.33.orig/x-setup	2006-07-19 16:55:42.000000000 -0400
+++ livecd-tools-1.0.33/x-setup	2006-08-01 15:33:10.000000000 -0400
@@ -38,12 +38,12 @@
 if [ -x /usr/sbin/opengl-update-livecd ]
 then
 	/usr/sbin/opengl-update-livecd $GLTYPE
-elif [ -x /usr/sbin/opengl-update ]
-then
-	/usr/sbin/opengl-update --dst-prefix=/etc/opengl $GLTYPE
 elif [ -x /usr/bin/eselect -a -e /usr/share/eselect/modules/opengl.eselect ]
 then
 	/usr/bin/eselect opengl set $GLTYPE --dst-prefix=/etc/opengl
+elif [ -x /usr/sbin/opengl-update ]
+then
+	/usr/sbin/opengl-update --dst-prefix=/etc/opengl $GLTYPE
 else
 	echo "ERROR: no opengl-update script can be located"
 	exit 1
